File: src/main/java/cz/cvut/fel/plichjan/ViewerFrame.java
package cz.cvut.fel.plichjan;

import cz.cvut.fel.plichjan.distmesh.DistMesh2D;
import cz.cvut.fel.plichjan.distmesh.result.Mesh;

import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.awt.geom.*;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.List;

/**
 * GUI class to test the delaunay_triangulation Triangulation package:
 */

public class ViewerFrame extends JFrame implements ActionListener, IViewer {

    private Thread dmThread;
    private List<Shape> shapes;
    private List<Shape> circles;
    private String fileName = "test.js";
    private Rectangle2D bbox;
    private float pixelSize;
    private AffineTransform transform;

    public static void main(String[] args) {
		ViewerFrame win = new ViewerFrame();
		win.start();
	}

	private static final long serialVersionUID = 1L;
	// *** private data ***
	public static final int DISMASH = 13;
	private int _stage;

    // *** text area ***
	public ViewerFrame() {
		this.setTitle("Delaunay GUI tester");
        addJPanel();
		this.setSize(500, 500);
		_stage = 0;

		addWindowListener(new WindowAdapter() {
            public void windowClosing(WindowEvent e) {
                System.exit(0);
            }
        });
	}

    private void addJPanel() {
        setLayout(new BorderLayout());
        final JPanel comp = new JPanel() {
            @Override
            protected void paintComponent(Graphics g) {
                super.paintComponent(g);
                Graphics2D g2 = (Graphics2D)g.create();
                if (bbox != null) {
                    applyLimits(g2, getWidth(), getHeight(), bbox, true);
                }
                ViewerFrame.this.paintComponent(g2);
            }
        };

        add(comp);
    }

    /**
     * Applies a coordinate transform to a Graphics2D graphics context.  The upper
     * left corner of the viewport where the graphics context draws is assumed to
     * be (0,0).  This method sets the global variables pixelSize and transform.
     *
     * @param g2 The drawing context whose transform will be set.
     * @param width The width of the viewport where g2 draws.
     * @param height The height of the viewport where g2 draws.
     * @param limitsRequested  Specifies a rectangle that will be visible in the
     *   viewport. Under the transform, the rectangle with corners (limitsRequested[0],
     *   limitsRequested[1]) and (limitsRequested[2],limitsRequested[3]) will just
     *   fit in the viewport.
     * @param preserveAspect if preserveAspect is false, then the limitsRequested
     *   rectangle will exactly fill the viewport; if it is true, then the limits
     *   will be expanded in one direction, horizontally or vertically, to make
     *   the aspect ratio of the displayed rectangle match the aspect ratio of the
     *   viewport.  Note that when preserveAspect is false, the units of measure in
     *   the horizontal and vertical directions will be different.
     */
    private void applyLimits(Graphics2D g2, int width, int height,
                             Rectangle2D limitsRequested, boolean preserveAspect) {
        Rectangle2D recLim = new Rectangle2D.Double();
        recLim.setRect(limitsRequested);
        if (preserveAspect) {
            double displayAspect = Math.abs((double) height / width);
            double requestedAspect = Math.abs(recLim.getHeight() / recLim.getWidth());
            if (displayAspect > requestedAspect) {
                double excess = recLim.getHeight() * (displayAspect / requestedAspect - 1);
                recLim.setRect(recLim.getX(), recLim.getY() - excess / 2, recLim.getWidth(), recLim.getHeight() + excess);
            } else if (displayAspect < requestedAspect) {
                double excess = recLim.getWidth() * (requestedAspect / displayAspect - 1);
                recLim.setRect(recLim.getX() - excess / 2, recLim.getY(), recLim.getWidth() + excess, recLim.getHeight());
            }
        }
        g2.translate(0, height);
        g2.scale(width / recLim.getWidth(), -height / recLim.getHeight());  // y grown up
        g2.translate(-recLim.getX(), -recLim.getY());
        double pixelWidth = Math.abs(recLim.getWidth() / width);
        double pixelHeight = Math.abs(recLim.getHeight() / height);
        pixelSize = (float) Math.min(pixelWidth, pixelHeight);
        transform = g2.getTransform();
    }

    public void paintComponent(Graphics2D g) {
        if (_stage == DISMASH) {
            drawDismash(g);
            if (circles != null) {
                drawMidpoints(g);
            }
        }
	}

    private void drawDismash(Graphics2D g) {
        g.setColor(Color.black);
        g.setStroke( new BasicStroke(pixelSize) );

        final ArrayList<Shape> list = new ArrayList<Shape>(shapes);
        for (Shape t : list) {
            g.draw(t);
        }
    }

    private void drawMidpoints(Graphics2D g) {
        g.setColor(Color.red);
        g.setStroke( new BasicStroke(pixelSize) );

        for (Shape t : circles) {
            g.draw(t);
        }
    }

	public void start() {
        createMenuBar();
        this.setVisible(true);
	}

	public void createMenuBar() {
		MenuBar mbar = new MenuBar();

		Menu m = new Menu("File");
		MenuItem m1;
		m1 = new MenuItem("Open");
		m1.addActionListener(this);
        m1.setShortcut(new MenuShortcut(KeyEvent.VK_O));
		m.add(m1);
		m1 = new MenuItem("Start DistMesh");
		m1.addActionListener(this);
		m1.setShortcut(new MenuShortcut(KeyEvent.VK_E));
		m.add(m1);
		MenuItem m6 = new MenuItem("Clear");
		m6.addActionListener(this);
        m6.setShortcut(new MenuShortcut(KeyEvent.VK_Q));
		m.add(m6);
		MenuItem m2 = new MenuItem("Exit");
		m2.addActionListener(this);
		m.add(m2);
		mbar.add(m);

		setMenuBar(mbar);
	}

	public void actionPerformed(ActionEvent evt) {
		String arg = evt.getActionCommand();
		if (arg.equals("Open")) {
            openTextFile();
        } else if (arg.equals("Start DistMesh")) {
            startDistMesh();
		} else if (arg.equals("Clear")) {
            _stage = 0;
            repaint();
		} else if (arg.equals("Exit")) {
			System.exit(209);
		}
	}

	// ********** Private methodes (open,save...) ********

	private void openTextFile() {
		_stage = 0;
		FileDialog d = new FileDialog(this, "Open text file", FileDialog.LOAD);
		d.setVisible(true);
		String dr = d.getDirectory();
		String fi = d.getFile();
        if (fi != null) { // the user actualy choose a file.
            fileName = dr + fi;
            startDistMesh();
		}
	}

    private void startDistMesh() {
        if (dmThread != null && dmThread.isAlive()) {
            return;
        }

        Runnable runnable = new Runnable() {
            public void run() {
                DistMesh2D distMesh2D = new DistMesh2D();
                distMesh2D.setViewer(ViewerFrame.this);

                ScriptEngineManager manager = new ScriptEngineManager();
                ScriptEngine engine = manager.getEngineByName("JavaScript");
                engine.put("distMesh2D", distMesh2D);
                engine.put("viewFrame", ViewerFrame.this);

                try {
                    engine.eval(new java.io.FileReader(fileName));
                } catch (ScriptException e) {
                    e.printStackTrace();
                } catch (FileNotFoundException e) {
                    e.printStackTrace();
                } finally {
                    System.err.format("%1$tT.%1$tL done.\n", Calendar.getInstance());
                }
            }
        };
        dmThread = new Thread(runnable, "DistMesh2D");
        System.err.format("%1$tT.%1$tL start.\n", Calendar.getInstance());
        dmThread.start();
    }

    @Override
    public void setNewPoints(List<Point2D> points, List<int[]> t, double[][] bbox) {
        _stage = DISMASH;
        this.bbox = new Rectangle2D.Double(bbox[0][0], bbox[0][1], 0, 0);
        this.bbox.add(bbox[1][0], bbox[1][1]);

        List<Shape> shapes = new ArrayList<Shape>(t.size());
        List<Shape> circles = new ArrayList<Shape>();
        for (int[] tr : t) {
            Point2D tmp;
            final Point2D a = points.get(tr[0]), b = points.get(tr[1]), c = points.get(tr[2]);

            if (tr.length == 3) {
                final Path2D path = new Path2D.Double(Path2D.WIND_NON_ZERO, 1+3);
                path.moveTo(a.getX(), a.getY());
                path.lineTo(b.getX(), b.getY());
                path.lineTo(c.getX(), c.getY());
                path.lineTo(a.getX(), a.getY());
                shapes.add(path);
            } else {
                final Path2D path = new Path2D.Double(Path2D.WIND_NON_ZERO, 1+3*2);
                path.moveTo(a.getX(), a.getY());
                tmp = getQuadPoint2D(a, b, points.get(tr[3]));
                path.quadTo(tmp.getX(), tmp.getY(), b.getX(), b.getY());
                tmp = getQuadPoint2D(b, c, points.get(tr[4]));
                path.quadTo(tmp.getX(), tmp.getY(), c.getX(), c.getY());
                tmp = getQuadPoint2D(c, a, points.get(tr[5]));
                path.quadTo(tmp.getX(), tmp.getY(), a.getX(), a.getY());
                shapes.add(path);

                addEllipse(circles, points.get(tr[3]));
                addEllipse(circles, points.get(tr[4]));
                addEllipse(circles, points.get(tr[5]));
            }
        }
        this.shapes = shapes;
        this.circles = circles;

        repaint();
    }

    private void addEllipse(List<Shape> shapes, Point2D a) {
        final double r = this.pixelSize * 5;
        shapes.add(new Ellipse2D.Double(a.getX() - r / 2, a.getY() - r / 2, r, r));
    }

    @Override
    public void drawMesh(Mesh mesh) {
        Rectangle2D.Double rec = null;
        List<Point2D> pdt = new ArrayList<Point2D>(mesh.getP().length);
        for (double[] point : mesh.getP()) {
            final Point2D.Double point2d = new Point2D.Double(point[0], point[1]);
            if (rec == null) {
                rec = new Rectangle2D.Double(point[0], point[1], 0, 0);
            }
            rec.add(point2d);
            pdt.add(point2d);
        }
        if (rec == null) {
            rec = new Rectangle2D.Double();
        }
        this.setNewPoints(pdt, Arrays.asList(mesh.getT()), new double[][]{{rec.getMinX(), rec.getMinY(),}, {rec.getMaxX(), rec.getMaxY(),}, });
    }

    private Point2D getQuadPoint2D(Point2D a, Point2D b, Point2D ab) {
        return new Point2D.Double(
                (4 * ab.getX() - a.getX() - b.getX()) / 2,
                (4 * ab.getY() - a.getY() - b.getY()) / 2
        );
    }

}

File: src/main/kotlin/cz/cvut/fel/plichjan/ViewerFrame.kt
package cz.cvut.fel.plichjan

import cz.cvut.fel.plichjan.distmesh.DistMesh2D
import cz.cvut.fel.plichjan.distmesh.result.Mesh
import java.awt.*
import java.awt.event.*
import java.awt.geom.*
import java.io.FileNotFoundException
import java.io.FileReader
import java.io.Serializable
import java.util.*
import javax.script.ScriptEngineManager
import javax.script.ScriptException
import javax.swing.*
import kotlin.math.abs
import kotlin.math.min

/**
 * GUI class to test the delaunay_triangulation Triangulation package:
 */
class ViewerFrame : JFrame(), ActionListener, IViewer, Serializable {

    private var dmThread: Thread? = null
    private var shapes: List<Shape>? = null
    private var circles: List<Shape>? = null
    private var fileName = "test.main.kts"
    private var bbox: Rectangle2D? = null
    private var pixelSize: Float = 0f
    private var transform: AffineTransform? = null

    private var _stage: Int = 0

    companion object {
        const val DISMASH = 13
        private const val serialVersionUID = 1L

        @JvmStatic
        fun main(args: Array<String>) {
            val win = ViewerFrame()
            win.start()
        }
    }

    init {
        title = "Delaunay GUI tester"
        addJPanel()
        setSize(500, 500)
        _stage = 0

        addWindowListener(object : WindowAdapter() {
            override fun windowClosing(e: WindowEvent) {
                System.exit(0)
            }
        })
    }

    private fun addJPanel() {
        layout = BorderLayout()
        val comp = object : JPanel() {
            override fun paintComponent(g: Graphics) {
                super.paintComponent(g)
                val g2 = g.create() as Graphics2D
                bbox?.let {
                    applyLimits(g2, width, height, it, true)
                }
                this@ViewerFrame.paintComponent(g2)
                g2.dispose()
            }
        }
        add(comp)
    }

    private fun applyLimits(
        g2: Graphics2D, width: Int, height: Int,
        limitsRequested: Rectangle2D, preserveAspect: Boolean
    ) {
        val recLim = Rectangle2D.Double()
        recLim.setRect(limitsRequested)
        if (preserveAspect) {
            val displayAspect = abs(height.toDouble() / width)
            val requestedAspect = abs(recLim.height / recLim.width)
            if (displayAspect > requestedAspect) {
                val excess = recLim.height * (displayAspect / requestedAspect - 1)
                recLim.setRect(recLim.x, recLim.y - excess / 2, recLim.width, recLim.height + excess)
            } else if (displayAspect < requestedAspect) {
                val excess = recLim.width * (requestedAspect / displayAspect - 1)
                recLim.setRect(recLim.x - excess / 2, recLim.y, recLim.width + excess, recLim.height)
            }
        }
        g2.translate(0, height)
        g2.scale(width / recLim.width, -height / recLim.height) // y grown up
        g2.translate(-recLim.x, -recLim.y)
        val pixelWidth = abs(recLim.width / width)
        val pixelHeight = abs(recLim.height / height)
        pixelSize = min(pixelWidth, pixelHeight).toFloat()
        transform = g2.transform
    }

    fun paintComponent(g: Graphics2D) {
        if (_stage == DISMASH) {
            drawDismash(g)
            if (circles != null) {
                drawMidpoints(g)
            }
        }
    }

    private fun drawDismash(g: Graphics2D) {
        g.color = Color.black
        g.stroke = BasicStroke(pixelSize)

        shapes?.let { currentShapes ->
            val list = ArrayList(currentShapes)
            for (t in list) {
                g.draw(t)
            }
        }
    }

    private fun drawMidpoints(g: Graphics2D) {
        g.color = Color.red
        g.stroke = BasicStroke(pixelSize)

        circles?.forEach { t ->
            g.draw(t)
        }
    }

    fun start() {
        createMenuBar()
        isVisible = true
    }

    private fun createMenuBar() {
        val mbar = MenuBar()

        val m = Menu("File")
        val m1 = MenuItem("Open").apply {
            addActionListener(this@ViewerFrame)
            shortcut = MenuShortcut(KeyEvent.VK_O)
        }
        m.add(m1)

        val m2 = MenuItem("Start DistMesh").apply {
            addActionListener(this@ViewerFrame)
            shortcut = MenuShortcut(KeyEvent.VK_E)
        }
        m.add(m2)

        val m6 = MenuItem("Clear").apply {
            addActionListener(this@ViewerFrame)
            shortcut = MenuShortcut(KeyEvent.VK_Q)
        }
        m.add(m6)

        val mExit = MenuItem("Exit").apply {
            addActionListener(this@ViewerFrame)
        }
        m.add(mExit)

        mbar.add(m)
        menuBar = mbar
    }

    override fun actionPerformed(evt: ActionEvent) {
        when (evt.actionCommand) {
            "Open" -> openTextFile()
            "Start DistMesh" -> startDistMesh()
            "Clear" -> {
                _stage = 0
                repaint()
            }
            "Exit" -> System.exit(209)
        }
    }

    private fun openTextFile() {
        _stage = 0
        val d = FileDialog(this, "Open text file", FileDialog.LOAD)
        d.isVisible = true
        val dr = d.directory
        val fi = d.file
        if (fi != null) {
            fileName = dr + fi
            startDistMesh()
        }
    }

    private fun startDistMesh() {
        if (dmThread?.isAlive == true) {
            return
        }

        val runnable = Runnable {
            val distMesh2D = DistMesh2D()
            distMesh2D.setViewer(this@ViewerFrame)

            // Populate ScriptContext for script support
            ScriptContext.distMesh2D = distMesh2D
            ScriptContext.viewFrame = this@ViewerFrame

            val manager = ScriptEngineManager()
            // Prefer kotlin script if possible, or fall back to javascript
            var engine = manager.getEngineByExtension("kts")
            if (engine == null) {
                engine = manager.getEngineByName("JavaScript")
            }
            
            if (engine == null) {
                System.err.println("No script engine found (tried kts, js)")
                return@Runnable
            }

            engine.put("distMesh2D_from_viewer", distMesh2D)
            engine.put("viewFrame", this@ViewerFrame)

            try {
                FileReader(fileName).use { reader ->
                    engine.eval(reader)
                }
            } catch (e: ScriptException) {
                e.printStackTrace()
            } catch (e: FileNotFoundException) {
                e.printStackTrace()
            } finally {
                System.err.format("%1\$tT.%1\$tL done.\n", Calendar.getInstance())
            }
        }
        dmThread = Thread(runnable, "DistMesh2D").apply {
            System.err.format("%1\$tT.%1\$tL start.\n", Calendar.getInstance())
            start()
        }
    }

    override fun setNewPoints(points: List<Point2D>, t: List<IntArray>, bbox: Array<DoubleArray>) {
        _stage = DISMASH
        val newBbox = Rectangle2D.Double(bbox[0][0], bbox[0][1], 0.0, 0.0)
        newBbox.add(bbox[1][0], bbox[1][1])
        this.bbox = newBbox

        val newShapes = ArrayList<Shape>(t.size)
        val newCircles = ArrayList<Shape>()
        for (tr in t) {
            val a = points[tr[0]]
            val b = points[tr[1]]
            val c = points[tr[2]]

            if (tr.size == 3) {
                val path = Path2D.Double(Path2D.WIND_NON_ZERO, 1 + 3)
                path.moveTo(a.x, a.y)
                path.lineTo(b.x, b.y)
                path.lineTo(c.x, c.y)
                path.lineTo(a.x, a.y)
                newShapes.add(path)
            } else {
                val path = Path2D.Double(Path2D.WIND_NON_ZERO, 1 + 3 * 2)
                path.moveTo(a.x, a.y)
                var tmp = getQuadPoint2D(a, b, points[tr[3]])
                path.quadTo(tmp.x, tmp.y, b.x, b.y)
                tmp = getQuadPoint2D(b, c, points[tr[4]])
                path.quadTo(tmp.x, tmp.y, c.x, c.y)
                tmp = getQuadPoint2D(c, a, points[tr[5]])
                path.quadTo(tmp.x, tmp.y, a.x, a.y)
                newShapes.add(path)

                addEllipse(newCircles, points[tr[3]])
                addEllipse(newCircles, points[tr[4]])
                addEllipse(newCircles, points[tr[5]])
            }
        }
        this.shapes = newShapes
        this.circles = newCircles

        repaint()
    }

    private fun addEllipse(shapes: MutableList<Shape>, a: Point2D) {
        val r = (this.pixelSize * 5).toDouble()
        shapes.add(Ellipse2D.Double(a.x - r / 2, a.y - r / 2, r, r))
    }

    override fun drawMesh(mesh: Mesh) {
        var rec: Rectangle2D.Double? = null
        val pdt = ArrayList<Point2D>(mesh.p!!.size)
        for (point in mesh.p!!) {
            val point2d = Point2D.Double(point[0], point[1])
            if (rec == null) {
                rec = Rectangle2D.Double(point[0], point[1], 0.0, 0.0)
            }
            rec.add(point2d)
            pdt.add(point2d)
        }
        if (rec == null) {
            rec = Rectangle2D.Double()
        }
        this.setNewPoints(
            pdt, mesh.t!!.toList(),
            arrayOf(doubleArrayOf(rec.minX, rec.minY), doubleArrayOf(rec.maxX, rec.maxY))
        )
    }

    private fun getQuadPoint2D(a: Point2D, b: Point2D, ab: Point2D): Point2D {
        return Point2D.Double(
            (4 * ab.x - a.x - b.x) / 2,
            (4 * ab.y - a.y - b.y) / 2
        )
    }
}
